### PR TITLE
Remove `ArchiveId.extension`

### LINKED ISSUE #568: `aqt list-qt`: Remove `--extension`/`--extensions` options
**Is your feature request related to a problem? Please describe.**
This is a UI/UX problem that I have with the `aqt list-qt` interface.

I think that the `--extension` flag should not be necessary to list architectures, modules, and archives. It's easy to forget that you need to use this flag. This flag is not necessary when using `install-qt` because the `install-qt` command is able to figure out which extension is necessary on its own. Why should `list-qt` be different?

If I am trying to list architectures using `aqt list-qt linux desktop 5.15.2`, I think it would be natural to expect `list-qt` to list every architecture available (gcc_64 and wasm_32). I do not expect to have to pass `--extension wasm` in order to find out that `wasm_32` is an available architecture.

**Describe the solution you'd like**
This is what I would like `aqt list-qt` to do when I use these CLI commands, without having to use the `--extension` flag:

```console
$ aqt list-qt linux desktop --arch 5.15.2
gcc_64 wasm_32
$ aqt list-qt linux android --arch 6.0.0
android_x86_64 android_x86 android_armv7 android_arm64_v8a
$ aqt list-qt linux android --modules 6.2.0 android_armv7
qt5compat qtquick3d qtquicktimeline qtshadertools ...
$ aqt list-qt linux android --archives 6.2.0 android_armv7
qtbase qtdeclarative qtquickcontrols2 qtsvg qttools qttranslations ...
```

All of the above commands currently either fail or report the wrong results, because the user has omitted the `--extension` flag.

**Describe alternatives you've considered**
I have tried using the existing interface. I think it's cumbersome and not fun to use.

**Additional context**
This will be an interface-breaking change. If anyone is using the `--extension` flag and parsing the output in an automated script, they will need to be warned about this.

This change will remove some functionality from `list-qt`. It is hard for me to imagine that any of our users are relying on these features, though.
* Currently, you can list all versions of Qt that have a particular architecture with:
  ```bash
  aqt list-qt <host> desktop --extension wasm     # show all versions that have the wasm_32 arch
  aqt list-qt <host> android --extension arm_v7   # show all versions that have the android_arm_v7 arch
  ```
  This capability will disappear, unless a new interface is added.
* Currently, you can list all preview versions of Qt available on download.qt.io, even though aqt is not capable of installing them:
  ```bash
  aqt list-qt linux desktop --extension preview
  aqt list-qt linux desktop --extension wasm_preview
  ```
  I don't see a need to keep this functionality, unless we plan to allow aqt to install preview versions.

### PR DESCRIPTION
**Pull request by ddalcino**
Fix #568

### CODE REVIEW

File: tests/test_list.py [hunk 1]
```diff
@@ -482,6 +467,72 @@ def _mock_fetch_http(_, rest_of_url, *args, **kwargs: str) -> str:
...
+        match = re.search(r"qt\d_\d+_(?P<arch>\w+)/Updates\.xml$", rest_of_url)
+        assert match
+        return xmltext.replace("armv7", match.group("arch"))
+
+    monkeypatch.setattr(MetadataFactory, "fetch_http", _mock_fetch_http)
+
+    # Unit test:
+    archive_id = ArchiveId("qt", host, target)
+    actual_arches = MetadataFactory(archive_id, architectures_ver=qt_ver_str).getList()
+    assert set(actual_arches) == expect_set
```
[the remark this answers is not in the document]
Comment (ddalcino) [reply tests/test_list.py:493]: fixed using a type hint

File: aqt/metadata.py [hunk 2]
```diff
@@ -490,6 +481,14 @@ def contains_qmake_exe(arch_path: Path) -> bool:
         selected_dir = QtRepoProperty.select_default_mingw(arch_dirs, is_dir=True)
         return installed_qt_version_dir / selected_dir if selected_dir else None
 
+    @staticmethod
+    def is_in_wasm_range(host: str, version: Version) -> bool:
+        return (
+            version in SimpleSpec(">=6.2.0")
```
[the remark this answers is not in the document]
Comment (ddalcino) [reply aqt/metadata.py:487]: incorrect

File: aqt/metadata.py [hunk 3]
```diff
@@ -205,66 +205,57 @@ class ArchiveId:
...
             major=qt_version_no_dots[0],
             ver=qt_version_no_dots,
-            ext="_" + self.extension if self.extension else "",
+            ext="_" + extension if extension else "",
         )
 
+    def all_extensions(self, version: Version) -> List[str]:
+        if self.target == "desktop" and QtRepoProperty.is_in_wasm_range(self.host, version):
+            return ["", "wasm"]
+        elif self.target == "android" and version >= Version("6.0.0"):
```
[the remark this answers is not in the document]
Comment (ddalcino) [reply aqt/metadata.py:249]: incorrect

File: aqt/metadata.py [hunk 4]
```diff
@@ -878,9 +830,10 @@ def fetch_archives_sde(self, cmd_type: str, version: Version) -> List[str]:
         assert (
             cmd_type in ("src", "doc", "examples") and self.archive_id.target == "desktop"
         ), "Internal misuse of fetch_archives_sde"
-        return self.fetch_archives(version, cmd_type, [])
+        return self.fetch_archives(version, cmd_type, [], is_sde=True)
 
-    def fetch_archives(self, version: Version, arch: str, modules: List[str]) -> List[str]:
+    def fetch_archives(self, version: Version, arch: str, modules: List[str], is_sde: bool = False) -> List[str]:
+        extension = "src_doc_examples" if is_sde else QtRepoProperty.extension_for_arch(arch, version >= Version("6.0.0"))
```
[the remark this answers is not in the document]
Comment (ddalcino) [reply aqt/metadata.py:836]: incorrect

File: aqt/metadata.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ddalcino) [reply aqt/metadata.py:487]: @sonatype-lift ignore

File: aqt/metadata.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ddalcino) [reply aqt/metadata.py:249]: @sonatype-lift ignore

File: aqt/metadata.py [hunk 5]
```diff
@@ -491,6 +482,14 @@ def contains_qmake_exe(arch_path: Path) -> bool:
         selected_dir = QtRepoProperty.select_default_mingw(arch_dirs, is_dir=True)
         return installed_qt_version_dir / selected_dir if selected_dir else None
 
+    @staticmethod
+    def is_in_wasm_range(host: str, version: Version) -> bool:
+        return (
+            version in SimpleSpec(">=6.2.0")
```
[the remark this answers is not in the document]
Comment (ddalcino) [reply aqt/metadata.py:488]: @sonatype-lift ignore

File: aqt/metadata.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ddalcino) [reply aqt/metadata.py:249]: @sonatype-lift ignore

**Comment by ddalcino** [comment]
For clarification, these are not new changes; I just rebased onto the current master to resolve any conflicts.

The sonatype-lift warnings here look like bugs in sonatype-lift, not aqtinstall:
* complaining that `SimpleSpec(">=6.2.0")` has too many arguments
* `version >= Version("6.0.0")` uses an unsupported operand

**Comment by miurahr** [comment]
I'd like to merge it to master.

**Pull request merged**